src/hazelweb/gui/ActionPanel.re

```diff
@@ -0,0 +1,464 @@
+
+  let info = text => {
+    info_button(true, text);
+  };
+
+  let info_action = (text, action) => {
+    info_button(is_action_allowed(action), text);
+  };
+
+  let is_action_allowed_with_on_type_check = (~on_type, action) => {
+    switch (cursor_info.typed) {
+    | OnType when on_type => is_action_allowed(action)
+    | OnType => false
+    | _ when on_type => false
+    | _ => is_action_allowed(action)
+    };
+  };
+
+  let combo_and_cursor = (~on_type, combo, description) => {
+    let is_allowed = is_action_allowed_with_on_type_check(~on_type);
+    combo_element(is_allowed, combo, description);
+  };
+
+  let operator_list = (~on_type, text, combos) => {
+    let actions =
+      List.map(
+        combo => {
+          let action = Hashtbl.find(Cell.kc_actions, combo, cursor_info);
+          (KeyCombo.get_details(combo), action);
+        },
+        combos,
+      );
+    let is_action_allowed = is_action_allowed_with_on_type_check(~on_type);
+    action_list(is_action_allowed, inject, actions, text);
+  };
+
+  let keyboard_button = combo => {
+    let action = Hashtbl.find(Cell.kc_actions, combo, cursor_info);
+    let combo = KeyCombo.get_details(combo);
+    keyboard_button(is_action_allowed, ~inject, ~combo, ~action);
+  };
+
+  let display_inline_block =
+    Vdom.Attr.(
+      style(Css_gen.(create(~field="display", ~value="inline-block")))
+    );
+
+  let spaced_line = children => {
+    action_label(
+      ~attrs=[
+        Vdom.Attr.(
+          style(
+            Css_gen.(
+              create(~field="display", ~value="flex")
+              @> create(~field="justify-content", ~value="space-between")
+            ),
+          )
+        ),
+      ],
+      children,
+    );
+  };
+
+  let dual_line = (description1, combo1, description2, combo2) => {
+    spaced_line([
+      Vdom.Node.div(
+        [display_inline_block],
+        [text(description1), keyboard_button(combo1)],
+      ),
+      Vdom.Node.div(
+        [display_inline_block],
+        [text(description2), keyboard_button(combo2)],
+      ),
+    ]);
+  };
+
+  let bottom_border_none =
+    Css_gen.create(~field="border-bottom", ~value="none");
+
+  let add_style = (style, elem) => {
+    switch (elem) {
+    | Vdom.Node.(Element(e)) =>
+      Vdom.Node.(Element(Element.add_style(e, style)))
+    | n => n
+    };
+  };
+
+  KeyCombo.[
+    section(
+      "Movement",
+      [
+        info([text("Move using arrow keys")]),
+        dual_line("Move to next hole ", Tab, "Previous hole ", ShiftTab),
+      ],
+    ),
+    section(
+      "General Editing",
+      [
+        spaced_line([
+          text("Backspace / Delete"),
+          Vdom.Node.div(
+            [display_inline_block],
+            [
+              keyboard_button(Backspace) |> add_style(bottom_border_none),
+              keyboard_button(Delete) |> add_style(bottom_border_none),
+            ],
+          ),
+        ]),
+        combo(Ctrl_Alt_I, simple("Swap expression up")),
```

Comment:
* this should be "swap line up/down"
* can we combine the up/down ones into a dual_line?

Reply:
I changed it to a the same style as `Backspace / Delete` as it looked a bit weird with the action boxes interspersed with the text. Now the whole UI is more uniform.